### DISCUSSION TITLE
Is it possible for a plugin to use the Address Field like Commerce v4 does in Store Settings?

### DISCUSSION THREAD

**Discussion by BenParizek** [craftcms/cms#11173] [Q&A] [answered]
I've tried to copy the code in the StoreController:

```
$locationField = Cp::addressCardsHtml(
            addresses: [],
            config: [
                'name' => 'locationAddressId',
                'maxAddresses' => 1,
            ]
);
```

https://github.com/craftcms/commerce/blob/77e3a1c0747574fb43ce54e598bdfabb87e789cf/src/controllers/StoreController.php#L35-L47

I can output the initial "Add an Address" UI but when I click on it I get an "User not authorized to create this element" error from the request to create the initial Address Element. On my first few passes, I don't see any way to change user permissions in the code.

Is it possible for our third-party plugins to use the Address Field in the same way that Craft Commerce does? If so, what stars do we need to align to get it to work without errors? Any tips or nudges to get me moving in the right direction would be appreciated.

_As a side note, your now first-party Address Field relies on the same library that I was using in the Sprout Address field. I'm planning on retiring the Sprout Address field as it feels silly to recreate the wheel when you folks have it in core now. If this is not possible and there is anything I can do on my end to help make this possible, I would rather spend my time helping make this use case possible than migrating an alternative option that nobody should need in a few months._

**Comment by BenParizek** [comment, marked as answer]
After a bit more digging, this appears possible.

----

I can create the field like so, passing it an address Element:

```
$locationField = Cp::addressCardsHtml(
  addresses: [$addressElement],
  config: [
      'name' => 'locationAddressId',
      'maxAddresses' => 1,
  ]
);
```

Return the code generated for the `addressCardsHtml` to the template and output it with the `|raw` filter:

```
{{ locationField|raw }}
```

And add an input to identify the address field I'm managing within the form the address exists:

{{ hiddenInput('locationAddressId', 7) }}

----

To address the permission issues on save, you will need to add support for the `Address::EVENT_AUTHORIZE_VIEW` and `Address::EVENT_AUTHORIZE_SAVE` events. See the example here in Commerce:

https://github.com/craftcms/commerce/blob/77e3a1c0747574fb43ce54e598bdfabb87e789cf/src/Plugin.php#L594-L611

And, as far as I can tell, the stuff in the [StoreLocationBehavior](https://github.com/craftcms/commerce/blob/develop/src/behaviors/StoreLocationBehavior.php) is handled in the above event and the behavior class never gets attached or used.

----

The Address Element gets stored when you save it from the Slideout editor. The "Label" field gets saved to `content.title` and the rest gets saved to the `address` table. The `locationAddressId` can be saved as you like to know where to find your data.

**Reply by brandonkelly** [reply]
Glad you figured it out!
